File: Student/controller.py
from flask import jsonify, request
from flask_jwt_extended import create_access_token, decode_token
from datetime import datetime
from Utilities import db, error_handler
from Student import validation
import hashlib


def add_quotes(data, value):
    if data == 'cgpa':
        return value
    else:
        return f'\'{value}\''


def get_user_from_token():
    token = str(request.headers['Authorization']).split(' ')[1]
    return decode_token(token)['sub']


def check_user():
    username = get_user_from_token()
    query = f"select count(*) from student where username = '{username}'"
    output = db.read_from_table(query)
    if output[0][0] == 1:
        return True
    else:
        return "user not found"


def give_hash(input):
    hash = hashlib.md5(input.encode())
    return hash.hexdigest()


def read_from_student():
    if isinstance(check_user(), bool):
        start_time = datetime.now()
        query = f"select * from student"
        output = db.read_from_table(query)
        return jsonify(
            db.give_response(data=output,
                             message='operation successful',
                             start_time=start_time))
    else:
        return jsonify(error_handler.generate_error_response(check_user()))


def read_by_id_student():
    if isinstance(check_user(), bool):
        id = request.args['id']
        start_time = datetime.now()
        validity = validation.validate_read(id)
        if isinstance(validity, bool):
            query = f"select * from student where id = '{id}'"
            output = db.read_from_table(query)
            return jsonify(
                db.give_response(data=output,
                                 message='operation successful',
                                 start_time=start_time))
        else:
            return jsonify(error_handler.generate_error_response(validity))
    else:
        return jsonify(error_handler.generate_error_response(check_user()))


def insert_in_student():
    if isinstance(check_user(), bool):
        start_time = datetime.now()
        input_data = {}
        input_data = request.get_json()
        validity = validation.validate_insert(input_data)
        if isinstance(validity, bool):
            query = f"insert into student values ('{input_data['id']}', '{input_data['name']}', '{input_data['department']}', {input_data['cgpa']}, '{input_data['username']}', '{input_data['password']}')"
            message = db.insert_into_table(query)
            return jsonify(
                db.give_response(data=[],
                                 message=message,
                                 start_time=start_time))
        else:
            return jsonify(error_handler.generate_error_response(validity))
    else:
        return jsonify(error_handler.generate_error_response(check_user()))


def update_in_student():
    if isinstance(check_user(), bool):
        start_time = datetime.now()
        input_data = {}
        input_data = request.get_json()
        validity = validation.validate_update(input_data)
        if isinstance(validity, bool):
            new_value = add_quotes(input_data['change_col'],
                                   input_data['new_value'])
            where_value = add_quotes(input_data['where_col'],
                                     input_data['where_value'])
            query = f"update student set {input_data['change_col']} = {new_value} where {input_data['where_col']} = {where_value}"
            message = db.update_table(query)
            return jsonify(
                db.give_response(data=[],
                                 message=message,
                                 start_time=start_time))
        else:
            return jsonify(error_handler.generate_error_response(validity))
    else:
        return jsonify(error_handler.generate_error_response(check_user()))


def delete_from_student():
    if isinstance(check_user(), bool):
        start_time = datetime.now()
        input_data = {}
        input_data = request.get_json()
        validity = validation.validate_delete(input_data)
        if isinstance(validity, bool):
            where_value = add_quotes(input_data['where_col'],
                                     input_data['where_value'])
            query = f"delete from student where {input_data['where_col']} = {where_value}"
            message = db.delete_from_table(query)
            return jsonify(
                db.give_response(data=[],
                                 message=message,
                                 start_time=start_time))
        else:
            return jsonify(error_handler.generate_error_response(validity))
    else:
        return jsonify(error_handler.generate_error_response(check_user()))


def student_login():

    start_time = datetime.now()
    input_data = {}
    input_data = request.get_json()
    validity = validation.validate_insert(input_data)

    if isinstance(validity, bool):
        query = f"select password from student where username = '{input_data['username']}'"
        store_password = db.read_from_table(query)

        if give_hash(input_data['password']) == store_password[0][0]:
            access_token = create_access_token(identity=input_data['username'])
            return jsonify(
                db.give_response(data=[],
                                 message="login successfull !!!",
                                 start_time=start_time,
                                 token=access_token))
        else:
            return jsonify(
                db.give_response(data=[],
                                 message="wrong username or password !!",
                                 start_time=start_time))
    else:
        return jsonify(error_handler.generate_error_response(validity))

File: Student/controller.js
const { executeQuery } = require("../Utilities/db");
const { makeResponse } = require("../Utilities/responseMaker");
const { encryptData } = require("../Utilities/encryptor");
const { tokenCreator } = require("./createToken");
const { errorObjectCreator } = require("../Utilities/errorHandler");
const jwt = require('jsonwebtoken');

// Default error code is set to 500 and it is changed whenever needed.
var errorCode = 500;

const readStudentData = async (req, res, next) => {
    try {
        let result = await executeQuery("select * from student").catch(
            function reject(error) {
                throw error;
            }
        );

        let response = makeResponse("query successful!!", result);
        return res.json(response);
    } catch (error) {
        let errorObject = errorObjectCreator(
            "error while reading student data",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const readStudentById = async (req, res, next) => {
    try {
        // extracting data from query parameters
        let data = req.query;
        const inputData = [data.id];
        let query = `select * from student where id = ?`;
        let result = await executeQuery(query, inputData).catch(function reject(error) {
            throw error;
        });

        // checking if object result is empty or not
        if (Object.keys(result).length == 0) {
            errorCode = 404;
            throw new Error("id not found");
        } else {
            return res.json(makeResponse("query successful!!", result));
        }
    } catch (error) {
        let errorObject = errorObjectCreator(
            "error while reading student data by id",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const studentSignup = async (req, res, next) => {
    try {
        let data = req.body;

        // encrypting the user entered password
        let encryptedPassword = encryptData(data.password);

        const inputData = [
            [
                [
                    data.id,
                    data.name,
                    data.department,
                    data.cgpa,
                    data.username,
                    encryptedPassword,
                    data.phone,
                ],
            ],
        ];

        let query = `insert into student(id, name, department, cgpa, username, password, phone) values ?`;
        const token = tokenCreator(data.username);
        let result = await executeQuery(query, inputData).catch(function reject(
            error
        ) {
            throw error;
        });

        return res.json(makeResponse(`Signup successfull your token = ${token}`));
    } catch (error) {
        let errorObject = errorObjectCreator(
            "error while signup",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const updateStudentData = async (req, res, next) => {
    try {
        let data = req.body;
        const inputData = [Object.values(data)[1]];
        let token = req.header("authorization").split(" ")[1];
        let username = getUser(token);
        
        let query = `update student set ${data.column} = ? where username = '${username}'`;

        let result = await executeQuery(query, inputData).catch(function reject(
            error
        ) {
            throw error;
        });

        // checking if any data was affected or not
        if (result.affectedRows == 0) {
            errorCode = 404;
            throw new Error("id not found");
        }
        return res.json(makeResponse(`your updated ${data.column} is ${Object.values(data)[1]}`));
    } catch (error) {
        errorCode = 400;
        let errorObject = errorObjectCreator(
            "error while updating student data",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const deleteStudentData = async (req, res, next) => {
    try {
        let token = req.header("authorization").split(" ")[1];
        let username = getUser(token);
        const inputData = [username];
        query = `delete from student where username = ?`;

        result = await executeQuery(query, inputData).catch(function reject(error) {
            throw error;
        });
        if (result.affectedRows == 0) {
            errorCode = 404;
            throw new Error("user not found");
        }
        return res.json(makeResponse(`records deleted for ${username}`));
    } catch (error) {
        let errorObject = errorObjectCreator(
            "error while deleting student data",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const studentLogin = async (req, res, next) => {
    try {
        let data = req.body;
        let username = data.username;
        const inputData = [data.username];

        // encrypting the user entered password
        encryptedPassword = encryptData(data.password);
        query = `select password from student where username = ?`;

        result = await executeQuery(query, inputData).catch(function reject(error) {
            throw error;
        });

        if (Object.keys(result).length == 0) {
            errorCode = 404;
            throw new Error("username not valid");
        } else {
            if (result[0].password == encryptedPassword) {
                const token = tokenCreator(username);
                return res.json(makeResponse(`login successfull !!! token = ${token}`));
            } else {
                errorCode = 406;
                throw new Error("wrong password !!!");
            }
        }
    } catch (error) {
        let errorObject = errorObjectCreator(
            "error while student login",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const insertStudentData = async (req, res, next) => {
    try {
        let data = req.body;

        const inputData = [];
        const dataArrayCollection = makeDataArray(data);

        inputData.push(dataArrayCollection);
        console.log(inputData);
        let query = `insert into student(id, name, department, cgpa, username, password, phone) values ?`;
        let result = await executeQuery(query, inputData).catch(function reject(
            error
        ) {
            throw error;
        });

        return res.json(makeResponse(result.affectedRows + " rows affected !!!"));
    } catch (error) {
        let errorObject = errorObjectCreator(
            "error while inserting student data",
            errorCode,
            error
        );
        next(errorObject);
    }
};

const makeDataArray = (data) => {
    const dataArrayCollection = [];
    for (let i = 0; i < Object.keys(data).length; i++) {
        //encrypting the user entered password
        tempData = data[i];
        let encryptedPassword = encryptData(tempData.password);

        const dataArray = [
            tempData.id,
            tempData.name,
            tempData.department,
            tempData.cgpa,
            tempData.username,
            encryptedPassword,
            tempData.phone,
        ];
        dataArrayCollection.push(dataArray);
    }
    return dataArrayCollection;
};

const getUser = (token) => {
    var decode = jwt.decode(token);
    return decode.username;
}

const checkUser = async (username) => {
        const inputData = [username];
        let query = `select * from student where username = ?`;
        let result = await executeQuery(query, inputData).catch(function reject(error) {
            throw error;
        });

        // checking if object result is empty or not
        if (Object.keys(result).length == 0) {
            return false;
        } else {
            return true;
        }
}

module.exports = {
    readStudentData,
    readStudentById,
    studentSignup,
    updateStudentData,
    deleteStudentData,
    studentLogin,
    insertStudentData,
    getUser,
    checkUser
}